plugins/base-gradle-plugin/src/main/groovy/org/kordamp/gradle/plugin/base/util/ExtensionUtil.java

```diff
@@ -0,0 +1,53 @@
+package org.kordamp.gradle.plugin.base.util;
+
+import org.gradle.api.Project;
+import org.gradle.api.plugins.ExtensionAware;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+
+public class ExtensionUtil {
+
+    public static <E extends ExtensionAware, P extends ExtensionAware> E createIfMissing(Project project, ExtensionPath<P, E> path) {
```

Comment:
I don't think you should have a `createIfMissing`. Instead, you should have a plugin which creates the extension. Then plugins which need the extension have to apply the plugin. If it's already applied, then it's not reapplied so it's guaranteed that the extension is present. If it's not yet applied, then the extension will be created.

e.g: https://github.com/micronaut-projects/micronaut-gradle-plugin/blob/master/src/main/java/io/micronaut/gradle/aot/MicronautAotPlugin.java#L117-L119

This creates an `aot` extension to the `micronaut` extension. The extension is itself created via the `base` plugin:

https://github.com/micronaut-projects/micronaut-gradle-plugin/blob/bbf99292e87ce026a5208ed0ca9d4163c73ec2fb/src/main/java/io/micronaut/gradle/MicronautBasePlugin.java#L24

This avoids convoluted logic about whether something is present or not: instead you declare your requirements: "I need the extension to be there"

Reply:
good point. I already know that the parent is present so it really makes no sense to check for presence.